### PR TITLE
Optimize `VectorizedInConstPredicate` using bitmap/array

### PR DESCRIPTION
**Pull request by dirtysalt**
ref: #1422 

I do a experiment on ssb_100g using 
- 8 scan threads and 
- set `parallel_fragment_exec_instance_num = 8`

Before improvement

```
mysql> select count(s_nation) from lineorder_flat where s_nation like '%N';
+-------------------+
| count(`s_nation`) |
+-------------------+
|          72048551 |
+-------------------+
1 row in set (0.92 sec)

```

And using this PR, time can be reduced for 26%

```
mysql> select count(s_nation) from lineorder_flat where s_nation like '%N';
+-------------------+
| count(`s_nation`) |
+-------------------+
|          72048551 |
+-------------------+
1 row in set (0.68 sec)
```

And if we enable global dict optimization `set cbo_enable_low_cardinality_optimize = true`, time can be reduces further for 52%.

```
mysql> select count(s_nation) from lineorder_flat where s_nation like '%N';
+-------------------+
| count(`s_nation`) |
+-------------------+
|          72048551 |
+-------------------+
1 row in set (0.44 sec)
```

### CODE REVIEW

File: be/src/storage/vectorized/column_predicate_rewriter.cpp [hunk 1]
```diff
@@ -336,8 +336,8 @@ bool ColumnPredicateRewriter::_rewrite_expr_predicate(ObjectPool* pool, const Co
     column_ref._type = type_desc;
     Expr* probe_expr = &column_ref;
     // probe_expr will be copied into filter, so we don't need to allocate it.
-    ExprContext* filter =
-            RuntimeFilterHelper::create_runtime_in_filter(state, pool, probe_expr, eq_null, null_in_set, is_not_in);
+    ExprContext* filter = RuntimeFilterHelper::create_runtime_in_filter(state, pool, probe_expr, eq_null, null_in_set,
```
Comment (kangkaisen) [inline be/src/storage/vectorized/column_predicate_rewriter.cpp:339]: shouldn't create runtime filter in column_predicate_rewriter.cpp

File: be/src/exprs/vectorized/in_const_predicate.hpp [hunk 2]
```diff
@@ -234,17 +305,54 @@ class VectorizedInConstPredicate final : public Predicate {
...
+        }
+    }
+
     const bool _is_not_in;
     bool _is_prepare;
     bool _null_in_set;
     bool _is_join_runtime_filter = false;
     bool _eq_null = false;
+    int _bitmap_size = 0;
+    std::vector<uint8_t> _bitmap_buffer;
```
Comment (kangkaisen) [inline be/src/exprs/vectorized/in_const_predicate.hpp:350]: Name to array_buffer

File: be/src/exprs/vectorized/in_const_predicate.hpp [hunk 3]
```diff
@@ -234,17 +305,54 @@ class VectorizedInConstPredicate final : public Predicate {
 
     void set_eq_null(bool value) { _eq_null = value; }
 
+    void set_bitmap_size(int bitmap_size) { _bitmap_size = bitmap_size; }
+
+    bool is_use_bitmap() const { return _bitmap_size != 0; }
+
 private:
+    // Note(yan): It's very tempting to use real bitmap. But the scenario is, the bitmap size is usually small like dict codes
```
Comment (kangkaisen) [inline be/src/exprs/vectorized/in_const_predicate.hpp:313]: Remove these comments.

**Comment by Seaven** [comment]
I think implement VectorizedInConstPredicate template special for INT type is better than now?

**Comment by dirtysalt** [comment]
> I think implement VectorizedInConstPredicate template special for INT type is better than now?

I wish. But to be honest, `VectorizedInConstPredicate` logic is quite complex enough considering (null-safe, is there is null in the set, and is-in or is-not-in). And this optimization is actually quite tightly coupled to `VectorizedInConstPredicate`, so better embedded into this class. 

Just imagine that what if there is a specialized `VectorizedInConstPredicate` class for INT, there must be a lot of duplicated code, which is quite hard to maintain.

**Comment by dirtysalt** [comment]
This PR has passed the daily run.

File: be/src/storage/vectorized/column_predicate_rewriter.cpp (hunk 1 above)
Comment (dirtysalt) [reply be/src/storage/vectorized/column_predicate_rewriter.cpp:339]: Yes, I'll refactor this piece of code and create a generic function to create `in-predicate`

**Comment by Seaven** [comment]
> > I think implement VectorizedInConstPredicate template special for INT type is better than now?
> 
> I wish. But to be honest, `VectorizedInConstPredicate` logic is quite complex enough considering (null-safe, is there is null in the set, and is-in or is-not-in). And this optimization is actually quite tightly coupled to `VectorizedInConstPredicate`, so better embedded into this class.
> 
> Just imagine that what if there is a specialized `VectorizedInConstPredicate` class for INT, there must be a lot of duplicated code, which is quite hard to maintain.

Yeah, you are right, but at the same time I think the modify will take the expr become more and more complicated.  We can abstract some method implement in prepare/init phase to simple the expr logic without use so many if-else

**Comment by dirtysalt** [comment]
> > > I think implement VectorizedInConstPredicate template special for INT type is better than now?
> > 
> > 
> > I wish. But to be honest, `VectorizedInConstPredicate` logic is quite complex enough considering (null-safe, is there is null in the set, and is-in or is-not-in). And this optimization is actually quite tightly coupled to `VectorizedInConstPredicate`, so better embedded into this class.
> > Just imagine that what if there is a specialized `VectorizedInConstPredicate` class for INT, there must be a lot of duplicated code, which is quite hard to maintain.
> 
> Yeah, you are right, but at the same time I think the modify will take the expr become more and more complicated. We can abstract some method implement in prepare/init phase to simple the expr logic without use so many if-else, such as we can packet a HashSet template with differet implement for different type

I'll follow your idea and do refactor.

**Comment by dirtysalt** [comment]
run starrocks_be_unittest

**Comment by dirtysalt** [comment]
run starrocks_clang-tidy

**Comment by dirtysalt** [comment]
passed daily test

**Review by kangkaisen — APPROVED** [review]

**Review by Seaven — APPROVED** [review]

**Pull request merged**